cloud/src/recycler/recycler.cpp

```diff
@@ -1888,72 +2019,90 @@ int InstanceRecycler::abort_timeout_txn() {
-                    .tag("key", hex(k))
-                    .tag("db_id", db_id)
-                    .tag("txn_id", txn_id);
-            return -1;
-        }
-        txn->put(recyc_txn_key, recyc_txn_val);
-        // Remove txn running key
-        txn->remove(k);
-        err = txn->commit();
-        if (err != TxnErrorCode::TXN_OK) {
-            LOG_WARNING("failed to commit txn err={}", err)
-                    .tag("key", hex(k))
-                    .tag("db_id", db_id)
-                    .tag("txn_id", txn_id);
-            return -1;
-        }
-        ++num_abort;
+        do {
+            std::unique_ptr<Transaction> txn;
+            TxnErrorCode err = txn_kv_->create_txn(&txn);
+            if (err != TxnErrorCode::TXN_OK) {
+                LOG_ERROR("failed to create txn err={}", err).tag("key", hex(k));
+                return -1;
+            }
+            std::string_view k1 = k;
+            //TxnRunningKeyInfo 0:instance_id  1:db_id  2:txn_id
+            k1.remove_prefix(1); // Remove key space
+            std::vector<std::tuple<std::variant<int64_t, std::string>, int, int>> out;
+            if (decode_key(&k1, &out) != 0) {
+                LOG_ERROR("failed to decode key").tag("key", hex(k));
+                return -1;
+            }
+            int64_t db_id = std::get<int64_t>(std::get<0>(out[3]));
+            int64_t txn_id = std::get<int64_t>(std::get<0>(out[4]));
+            VLOG_DEBUG << "instance_id=" << instance_id_ << " db_id=" << db_id
+                       << " txn_id=" << txn_id;
+            // Update txn_info
+            std::string txn_inf_key, txn_inf_val;
+            txn_info_key({instance_id_, db_id, txn_id}, &txn_inf_key);
+            err = txn->get(txn_inf_key, &txn_inf_val);
+            if (err != TxnErrorCode::TXN_OK) {
+                LOG_WARNING("failed to get txn info err={}", err).tag("key", hex(txn_inf_key));
+                return -1;
+            }
+            TxnInfoPB txn_info;
+            if (!txn_info.ParseFromString(txn_inf_val)) {
+                LOG_WARNING("failed to parse txn info").tag("key", hex(k));
+                return -1;
+            }
+
+            if (TxnStatusPB::TXN_STATUS_COMMITTED == txn_info.status()) {
+                txn.reset();
+                std::shared_ptr<TxnLazyCommitTask> task =
+                        txn_lazy_committer_->submit(instance_id_, txn_info.txn_id());
+                std::pair<MetaServiceCode, std::string> ret = task->wait();
+                if (ret.first != MetaServiceCode::OK) {
+                    LOG(WARNING) << "lazy commit txn failed txn_id=" << txn_id
+                                 << " code=" << ret.first << "msg=" << ret.second;
+                    return -1;
+                }
+                continue;
+            } else {
+                DCHECK(txn_info.status() != TxnStatusPB::TXN_STATUS_VISIBLE);
+                txn_info.set_status(TxnStatusPB::TXN_STATUS_ABORTED);
```

Comment:
The `continue` means retry to recycle this "timeouted" TXN, but it has been committed (status visible) via the `TxnLazyCommitTask`, then this DCHECK will panic.